File: CS/UserDiffsToDB.Module.Win/ImportExportUserDifferencesController.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;
using System.Windows.Forms;
using System.IO;
using DevExpress.ExpressApp.Core;
using DevExpress.ExpressApp.Model.Core;
using DevExpress.ExpressApp.Security;
using DevExpress.Persistent.BaseImpl;

namespace UserDiffsToDB.Module {
    public class ImportExportUserDifferencesController : WindowController {
        private SimpleAction exportDifferencesAction;
        private SimpleAction importDifferencesAction;
        public ImportExportUserDifferencesController() {
            exportDifferencesAction = new SimpleAction(this, "ExportUserDifferences", PredefinedCategory.Tools);
            exportDifferencesAction.ImageName = "Action_LocalizationExport";
            exportDifferencesAction.Execute += exportDifferencesAction_Execute;
            importDifferencesAction = new SimpleAction(this, "ImportUserDifferences", PredefinedCategory.Tools);
            importDifferencesAction.ImageName = "Action_LocalizationImport";
            importDifferencesAction.Execute += importDifferencesAction_Execute;
            this.TargetWindowType = WindowType.Main;
        }
        public void UpdateActivity() {
            bool isAdministrator = ((SimpleUser)SecuritySystem.CurrentUser).IsAdministrator;
            importDifferencesAction.Active["Security"] = isAdministrator;
            exportDifferencesAction.Active["Security"] = isAdministrator;
        }
        private void exportDifferencesAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            SaveFileDialog saveFileDialog = new SaveFileDialog();
            saveFileDialog.AddExtension = true;
            saveFileDialog.Filter = "Model differences files (*.xafml)|*.xafml";
            saveFileDialog.FileName = ModelDifferenceStore.UserDiffDefaultName + ".xafml";
            if(saveFileDialog.ShowDialog(Form.ActiveForm) == DialogResult.OK) {
                string file = Path.GetFileNameWithoutExtension(saveFileDialog.FileName);
                string path = Path.GetDirectoryName(saveFileDialog.FileName);
                FileModelStore fileModelStore = new FileModelStore(path, file);
                Frame.SynchronizeInfo();
                fileModelStore.SaveDifference(((ModelApplicationBase)Application.Model).LastLayer);
            }
        }
        private void importDifferencesAction_Execute(object sender, SimpleActionExecuteEventArgs e) {
            OpenFileDialog openFileDialog = new OpenFileDialog();
            openFileDialog.AddExtension = true;
            openFileDialog.Filter = "Model differences files (*.xafml)|*.xafml";
            openFileDialog.FileName = ModelDifferenceStore.UserDiffDefaultName + ".xafml";
            if(openFileDialog.ShowDialog(Form.ActiveForm) == DialogResult.OK) {
                string file = Path.GetFileNameWithoutExtension(openFileDialog.FileName);
                string path = Path.GetDirectoryName(openFileDialog.FileName);
                FileModelStore fileModelStore = new FileModelStore(path, file);
                ApplicationModelsManager.RereadLastLayer(fileModelStore, Application.Model);
                Frame.View.SynchronizeWithInfo();
                UpdateActivity();
            }
        }
        protected override void OnWindowChanging(Window window) {
            base.OnWindowChanging(window);
            UpdateActivity();
        }
    }
}

File: VB/UserDiffsToDB.Module.Win/ImportExportUserDifferencesController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base
Imports System.Windows.Forms
Imports System.IO
Imports DevExpress.ExpressApp.Core
Imports DevExpress.ExpressApp.Model.Core
Imports DevExpress.ExpressApp.Security
Imports DevExpress.Persistent.BaseImpl

Namespace UserDiffsToDB.Module
    Public Class ImportExportUserDifferencesController
        Inherits WindowController
        Private exportDifferencesAction As SimpleAction
        Private importDifferencesAction As SimpleAction
        Public Sub New()
            exportDifferencesAction = New SimpleAction(Me, "ExportUserDifferences", PredefinedCategory.Tools)
            exportDifferencesAction.ImageName = "Action_LocalizationExport"
            AddHandler exportDifferencesAction.Execute, AddressOf exportDifferencesAction_Execute
            importDifferencesAction = New SimpleAction(Me, "ImportUserDifferences", PredefinedCategory.Tools)
            importDifferencesAction.ImageName = "Action_LocalizationImport"
            AddHandler importDifferencesAction.Execute, AddressOf importDifferencesAction_Execute
            Me.TargetWindowType = WindowType.Main
        End Sub
        Public Sub UpdateActivity()
            Dim isAdministrator As Boolean = (CType(SecuritySystem.CurrentUser, SimpleUser)).IsAdministrator
            importDifferencesAction.Active("Security") = isAdministrator
            exportDifferencesAction.Active("Security") = isAdministrator
        End Sub
        Private Sub exportDifferencesAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim saveFileDialog As New SaveFileDialog()
            saveFileDialog.AddExtension = True
            saveFileDialog.Filter = "Model differences files (*.xafml)|*.xafml"
            saveFileDialog.FileName = ModelDifferenceStore.UserDiffDefaultName & ".xafml"
            If saveFileDialog.ShowDialog(Form.ActiveForm) = DialogResult.OK Then
                Dim file As String = System.IO.Path.GetFileNameWithoutExtension(saveFileDialog.FileName)
                Dim path As String = System.IO.Path.GetDirectoryName(saveFileDialog.FileName)
                Dim fileModelStore As New FileModelStore(path, file)
                Frame.SaveModel()
                fileModelStore.SaveDifference((CType(Application.Model, ModelApplicationBase)).LastLayer)
            End If
        End Sub
        Private Sub importDifferencesAction_Execute(ByVal sender As Object, ByVal e As SimpleActionExecuteEventArgs)
            Dim openFileDialog As New OpenFileDialog()
            openFileDialog.AddExtension = True
            openFileDialog.Filter = "Model differences files (*.xafml)|*.xafml"
            openFileDialog.FileName = ModelDifferenceStore.UserDiffDefaultName & ".xafml"
            If openFileDialog.ShowDialog(Form.ActiveForm) = DialogResult.OK Then
                Dim file As String = System.IO.Path.GetFileNameWithoutExtension(openFileDialog.FileName)
                Dim path As String = System.IO.Path.GetDirectoryName(openFileDialog.FileName)
                Dim fileModelStore As New FileModelStore(path, file)
                ApplicationModelsManager.RereadLastLayer(fileModelStore, Application.Model)
                Frame.View.LoadModel()
                UpdateActivity()
            End If
        End Sub
        Protected Overrides Sub OnWindowChanging(ByVal window As Window)
            MyBase.OnWindowChanging(window)
            UpdateActivity()
        End Sub
    End Class
End Namespace
